text/0000-ember-data-improved-ajax.md

```diff
@@ -0,0 +1,207 @@
+asynchronous configurations (e.g. pre-flight request to refresh an
+authentication token), `_requestFor` returns a promise which resolves with an
+object describing the request to make. The `_requestFor` method is private and
+calls the new public hooks.
+
+``` js
+_requestFor(params) {
+  let method  = this.methodForRequest(params);
+  let url     = this.urlForRequest(params);
+  let headers = this.headersForRequest(params);
+  let body    = this.bodyForRequest(params);
+
+  return RSVP.hash({ method, url, headers, body });
+}
+```
+##### `methodForRequest`
+
+Invoked to get the HTTP method for the request. Should return one of `"GET"`,
+`"PUT"`, `"POST"` or `"DELETE"`.
+
+##### `urlForRequest`
+
+Get the URL for the request. This hook returns the URL for the request which
+should be made, e.g. `/users/1` or `api.example.com:4321/users?name=abc`
+
+This hook uses 2 new public sub-hooks to construct the URL:
+
+- `pathForRequest` - re-uses the existing `urlForXXX` hooks
+- `queryForRequest` - get an object for all query params for the request
+
+The existing [`sortQueryParams`](http://emberjs.com/api/data/classes/DS.RESTAdapter.html#method_sortQueryParams)
+hook is reused within the new `queryForRequest`. An warning/deprecation should
+be logged when the path already contains query params, as the new
+`queryForRequest` hook should be used for that.
+
+Pseudo implementation:
+
+``` js
+urlForRequest(options) {
+  let path = this.pathForRequest(options);
+  let query = this.queryForRequest(options);
+
+  if (path.indexOf('?') !== -1) {
+    warning("use queryForRequest to specify query params");
+  } else if (query) {
+    let sortedQueryParams = this.sortQueryParams(query);
+    let queryParams = Ember.$.param(sortedQueryParams);
+
+    return `${path}?${queryParams}`;
+  }
+
+  return path;
+},
+
+pathForRequest(options) {
+  // re-use existing urlForXXX hooks
+},
+
+queryForRequest({ requestType, query }) {
+  if (requestType === "query") {
+    return query;
+  }
+}
+```
+
+##### `bodyForRequest` (currently `dataForRequest` in `ds-improved-ajax`)
+
+Body for the HTTP request which should be made.
+
+**Note**: in the current implementation of `ds-improved-ajax` the
+`bodyForRequest` hook is called `dataForRequest`. To be more agnostic of the
+underlying library used for making the request, it has been renamed to
+`bodyForRequest` within this RFC.
+
+##### `headersForRequest`
+
+All HTTP headers for this request. By default it returns the value for the
+`headers` computed property which can be set on the adapter. This hooks allows
+to add custom headers for each request individually.
+
+### `makeRequest`
+
+The public `makeRequest` method is invoked with all the properties that
+describe the request which should be made.
+
+``` js
+/**
+  @method
+  @public
+  @param request Object containing all properties for the request which should be made
+  @return Promise
+ */
+makeRequest(request) {
```

Comment:
I think we should enumerate all of the properties Ember Data expects/supports on the `request` object.